### PR TITLE
[REVIEW] Update to use type_dispatcher

### LINKED ISSUE #393: Update operators to use the `type_dispatcher`
Soon, PR #379 will be merged (one way or another), and we will be able to use templates instead of switch statements to implement operators taking type-erased `gdf_column`'s.

I suggest we use this issue to 'hang' commits which make this transition for different operators, and close it when they're all done.

I'll begin with what the locations where I see such a transition being appropriate:

* ~~`src/reductions.cu`:~~ fixed by https://github.com/rapidsai/cudf/pull/445
 *  ~~All operators defined using the switch-based `DEF_REDUCE_OP_NUM()` macro (and the macro itself).~~
  * ~~All operators defined using the switch-based `DEF_REDUCE_OP_REAL()` macro (and the macro itself).~~
* `src/io/convert/gdf_to_csr.cu`:
  * `gdf_to_csr()`
  * the, `convertDataElement()` <- Try to get rid of this at some point.
* `src/sqlsops.cu`:
  * ~~The various `gdf_group_by_` operators.~~ obviated by https://github.com/rapidsai/cudf/pull/426

https://github.com/rapidsai/cudf/blob/master/cpp/src/io/csv/csv_reader.cu#L164

https://github.com/rapidsai/cudf/blob/master/cpp/src/io/csv/csv_reader.cu#L764

https://github.com/rapidsai/cudf/blob/master/cpp/src/io/csv/csv_reader.cu#L1056

https://github.com/rapidsai/cudf/blob/c9fbc8b275315d173e4345e10283b6c697fa7556/cpp/src/quantiles/quantiles.cu#L156

https://github.com/rapidsai/cudf/blob/c9fbc8b275315d173e4345e10283b6c697fa7556/cpp/src/sort/segmented_sorting.cu#L241

https://github.com/rapidsai/cudf/blob/c9fbc8b275315d173e4345e10283b6c697fa7556/cpp/src/join/joining.cu#L147

https://github.com/rapidsai/cudf/blob/c9fbc8b275315d173e4345e10283b6c697fa7556/cpp/src/unary/unary_ops.cu#L94 & below 

https://github.com/rapidsai/cudf/blob/c9fbc8b275315d173e4345e10283b6c697fa7556/cpp/tests/utilities/cudf_test_utils.cuh#L78

https://github.com/rapidsai/cudf/blob/branch-0.5/cpp/tests/filter/helper/utils.cuh#L19

https://github.com/rapidsai/cudf/blob/b7356d9204acc5922a54fd7814c105390aaaafa5/cpp/src/filter/filter_ops.cu#L258

  * TO BE CONTINUED

### PR DESCRIPTION
**Pull request by devavret**
Closes #393 

Updated `switch` statements to `type_dispatcher` in

- [x] Join
- [x] Quantiles
- [x] Filter Ops
- [x] Segmented Sort
- [x] Radix Sort
- [x] Hash Groupby

Added a type trait `wrapped_type` and its helper `wrapped_type_t` to get the underlying type for `wrapper` object (`date32`, `date64`, `timestamp`, `category`)

Usage:
```c++
// when T = date32
using Tu = cudf::detail::unwrapped_type_t<T>;
// Tu = int

// when T = long int
using Tu = cudf::detail::unwrapped_type_t<T>;
// Tu = long int
```

Overloaded `std::numeric_limits` for `wrapper_types`. Currently has `max`, `min` and `lowest`.

### CODE REVIEW

File: cpp/src/groupby/hash_groupby.cuh [hunk 1]
```diff
@@ -80,6 +82,23 @@ struct is_same_functor : std::false_type{};
...
+ *---------------------------------------------------------------------------**/
+template <template <typename T> class op>
+struct typed_groupby_functor
+{
+  template <typename TypeAgg, typename... Ts>
+  gdf_error operator()(Ts&&... args)
+  {
+    // TypeAgg can a wrapped type (e.g detail::wrapper<gdf_date32, GDF_DATE32>)
+    // in which case, we want to pass the underlying type (gdf_date32 = int)
+    using TypeAggUndl = cudf::detail::unwrapped_type_t<TypeAgg>;
```
Comment (jrhemstad) [inline cpp/src/groupby/hash_groupby.cuh]: Why is this necessary? 

Ideally you shouldn't need to access the underyling type of the wrapper. That's considered to be "hidden" information.

Comment (devavret) [reply cpp/src/groupby/hash_groupby.cuh]: I tried but it broke at a call to `atomicCAS`.

Comment (jrhemstad) [reply cpp/src/groupby/hash_groupby.cuh]: In that case, what we'll need to do is to add some overloads for `atomicCAS` for the wrapper types.

Inside the `atomicCAS` overloads, you can access the underlying `value` and pass it through the "normal" `atomicCAS`.

I believe @kovaltan is doing something similar in https://github.com/rapidsai/cudf/pull/1005 and you could coordinate that effort with him.

Comment (kovaltan) [reply cpp/src/groupby/hash_groupby.cuh]: Sorry, I also use underlying type of the wrapper #1005 
I think using  underlying type  for cuda kernels is much simpler.

https://github.com/rapidsai/cudf/blob/e68dc4353db5c774b0c3a9bc651a4f4822910306/cpp/src/reductions/reductions.cu#L112-L119

Comment (jrhemstad) [reply cpp/src/groupby/hash_groupby.cuh]: We really don't want to do that, the underlying type should not be used directly here. It should only be used inside `wrapper` operators and specializations of `atomic*` operations.

The reason for this is because some of the wrapper types may have different semantics than their underlying type. Perfect example is `cudf::bool8` uses a `int8_t` as the underlying type. However `cudf::bool8{1} + cudf::bool8{1} == 1`, if you were to use the underlying type directly, you would produce the incorrect results of `1 + 1 = 2`.

You should think of the `wrapper` types as opaque objects that you should not be accessing the underlying type nor value.

The only entities that should know about the underlying type or value are `wrapper` operators and other operators that are specialized to work on the `wrapper` types, e.g., `atomic*` operations.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: cpp/src/groupby/hash_groupby.cuh (hunk 1 above)
Comment (jrhemstad) [reply cpp/src/groupby/hash_groupby.cuh]: To be clear, I'm highly averse to this trait. It encourages treating the `wrapper` types as something other than opaque types. 

This is the same reason that the use of `unwrap` is highly discouraged.

Comment (devavret) [reply cpp/src/groupby/hash_groupby.cuh]: Thanks to @kovaltan's atomics, I was able to refactor `hash_groupby` without using the trait. But to fix that, I had to make a clone of `std::numeric_limits` that works on wrapper types. That, however, needs that trait. I agree that this trait should not be used by functionality code and that's why it's inside the `detail` namespace. Ideally I intend this to be useful to utility writers.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 ## Improvements
 
+- PR #1076 Use `type_dispatcher` in join, quantiles, filter, segmented sort, radix sort and hash_groupby
+- PR #1076 Added type trait to get underlying type of wrapped type
```
Comment (harrism) [inline CHANGELOG.md]: Please don't add two lines for the same PR. I don't think the changelog has to list every single change you made, so you can delete this second line.

File: cpp/src/hash/helper_functions.cuh [hunk 3]
```diff
@@ -48,41 +48,11 @@ __inline__ __device__ int16_t atomicCAS(int16_t* address, int16_t compare, int16
...
-{
-  return (int64_t)atomicCAS((unsigned long long*)address, (unsigned long long)compare, (unsigned long long)val);
-}
-
-__inline__ __device__ uint64_t atomicCAS(uint64_t* address, uint64_t compare, uint64_t val)
-{
-  return (uint64_t)atomicCAS((unsigned long long*)address, (unsigned long long)compare, (unsigned long long)val);
-}
-
 __inline__ __device__ long long int atomicCAS(long long int* address, long long int compare, long long int val)
```
Comment (harrism) [inline cpp/src/hash/helper_functions.cuh]: Wait, why are there still atomicCAS functions in this file? Shouldn't they all have been incorporated in cpp/src/utilities/device_atomics.cuh?

File: cpp/src/utilities/wrapper_types.hpp [hunk 4]
```diff
@@ -291,8 +292,78 @@ CUDA_HOST_DEVICE_CALLABLE
...
+
 } // namespace detail
 
+/**---------------------------------------------------------------------------*
+ * @brief Clone of std::numeric_limits that also works on wrapper types
+ * 
+ * @tparam T 
+ *---------------------------------------------------------------------------**/
+template <typename T>
+struct numeric_limits {
```
Comment (harrism) [inline cpp/src/utilities/wrapper_types.hpp]: If I understand correctly, an alternative approach would be to provide a specialization of `std::numeric_limits` for our wrapper types rather than providing a new numeric_limits class...  There is debate about whether doing so is legal in C++11, but it seems to me that it is legal to specialize `std::numeric_limits` for non-standard types according to this post: https://stackoverflow.com/questions/16122912/is-it-ok-to-specialize-stdnumeric-limitst-for-user-defined-number-like-class

Did you consider this approach? It would avoid violating the opaqueness of wrapper types that @jrhemstad wants to preserve, instead defining semantics for min() and max() of those wrapper types.

File: cpp/src/hash/helper_functions.cuh (hunk 3 above)
Comment (devavret) [reply cpp/src/hash/helper_functions.cuh]: Oh yeah, about that, I found that my compiler complained about the equivalence of `long long` and `long long int` (it thinks they're different) and the overload in `device_atomics.cuh` uses `int64_t` which was set to `long long` in my system headers. I would suggest making the current overloads templated like so 
```c++
template <typename T>
__forceinline__ __device__
T atomicCAS(T* address, T compare, T val)
{
    return cudf::detail::typesAtomicCASImpl<T, sizeof(T)>()(address, compare, val);
}
```
but I didn't include them here because it will be a pain for @kovaltan while merging #1308.
As for the preceding `atomicCAS(int8_t)`, it can be removed once #1308 is merged.

File: cpp/src/utilities/wrapper_types.hpp (hunk 4 above)
Comment (devavret) [reply cpp/src/utilities/wrapper_types.hpp]: I did consider and saw the same link you posted but considering that the other answer advised against this approach, I thought I'm not aware enough about the standards to make a call, so I chose the safer option. Although that option is quite appealing.

Comment (devavret) [reply cpp/src/utilities/wrapper_types.hpp]: ok, so I took your advice https://github.com/rapidsai/cudf/pull/1076/commits/d18f8dc9d459b0aaf993e06a69b39f5672d4b4aa and nothing seems to break.

File: cpp/src/utilities/wrapper_types.hpp [hunk 5]
```diff
@@ -305,4 +348,45 @@ using date64 = detail::wrapper<gdf_date64, GDF_DATE64>;
...
+template <typename T, gdf_dtype type_id>
+struct numeric_limits< cudf::detail::wrapper<T, type_id> > {
+  
+  using wrapper_t = cudf::detail::wrapper<T, type_id>;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the maximum finite value representable by the numeric type T
+   *---------------------------------------------------------------------------**/
+  static constexpr wrapper_t max() noexcept {
+    return wrapper_t{ std::numeric_limits<T>::max() };
```
Comment (harrism) [inline cpp/src/utilities/wrapper_types.hpp]: So you are still basically accessing the underlying type of the wrapper. What if we want to provide different semantics for `max()` and `lowest()` for the wrapper types than for the underlying type?

File: cpp/src/hash/helper_functions.cuh (hunk 3 above)
Comment (harrism) [reply cpp/src/hash/helper_functions.cuh]: My question is more about why there are atomic functions in two places. Why can't they all go in cpp/src/utilities/device_atomics.cuh?

Comment (harrism) [reply cpp/src/hash/helper_functions.cuh]: Also, to avoid losing your suggestions, please file an issue.

**Comment by harrism** [comment]
rerun tests

File: cpp/src/utilities/wrapper_types.hpp (hunk 5 above)
Comment (devavret) [reply cpp/src/utilities/wrapper_types.hpp]: Then you'd specialize this template to handle each different wrapper type. Something on the lines of:
```c++
template <>
struct numeric_limits< cudf::date32 > {
  using wrapper_t = cudf::date32;
  static constexpr wrapper_t max() noexcept { ... }
  static constexpr wrapper_t lowest() noexcept { ... }
  static constexpr wrapper_t min() noexcept { ... }
};
```
That wasn't the case for the wrapper types that currently exist though. I checked for dates and their mix and max are indeed the min and max of the integer types of their underlying data. You might think this wouldn't be the case for category types but in a discussion yesterday with @kkraus14 and @felipeblazing , I found out it *was* the case. (They said that comparison operators (>,<,>=,<=,==) should work, which I took to assume that min and max are also defined in the same way)

File: cpp/src/hash/helper_functions.cuh (hunk 3 above)
Comment (devavret) [reply cpp/src/hash/helper_functions.cuh]: >Why can't they all go in cpp/src/utilities/device_atomics.cuh?

They can and they should. I'll do that if #1308 gets merged first. Or I'll let @kovaltan know because if this PR gets merged first, he'll have a ODR violation compilation issue.

Comment (kovaltan) [reply cpp/src/hash/helper_functions.cuh]: PR #1308 just got merged. Please try it.

Comment (devavret) [reply cpp/src/hash/helper_functions.cuh]: Great, 👏🏼 now **I** have the ODR violation.

File: cpp/src/sort/segmented_sorting.cu [hunk 6]
```diff
@@ -192,70 +193,64 @@ gdf_error gdf_segmented_radixsort_plan_free(gdf_segmented_radixsort_plan_type *h
 }
 
 
+template <typename Tv>
+struct gdf_segmented_radixsort_functor
+{
+    template <typename Tk>
+    typename std::enable_if_t<!std::is_arithmetic<Tk>::value, gdf_error>
```
[the remark this answers is not in the document]
Comment (devavret) [reply cpp/src/sort/segmented_sorting.cu]: I ran into an issue here. The sort here is implemented with cub and cub has its [own traits](https://github.com/rapidsai/thirdparty-cub/blob/b165e1fb11eeea64ccf95053e40f2424312599cc/cub/util_type.cuh#L1104-L1124) that it uses to deal with keys. So there's a compilation error for wrapped types.

**Comment by harrism** [comment]
Thanks @devavret, great work. Finally merging!
